### PR TITLE
[0.2] Manage customer groups

### PR DESCRIPTION
**Pull request by adam-code-labx**
This PR will soon include a new section under settings to manage translatable groups.

Initially includes migration from text to json and model casting.
Extracted as a trait to reuse this for future migrations converting field from text to json.

Included test for the trait which covers the usual exceptions and validates migration and rollback.
@glennjacobs hoping this will also work well with PostgreSQL🤞

### CODE REVIEW

**Comment by glennjacobs** [comment]
- [ ] Need to decide if customer group names are to be changed to translatable and update accordingly
- [ ] Need a delete facility

**Comment by glennjacobs** [comment]
@alecritson need to add in the facility to edit the default customer group

**Comment by alecritson** [comment]
> @alecritson need to add in the facility to edit the default customer group

Should be in

File: packages/admin/resources/views/livewire/components/customers/show.blade.php [hunk 1]
```diff
@@ -67,7 +67,7 @@
                                        wire:key="group_{{ $group->id }}">
                                     <x-hub::input.checkbox wire:model.debounce.500ms="syncedGroups"
                                                            value="{{ $group->id }}" /> <span
-                                          class="ml-2">{{ $group->name }}</span>
+                                          class="ml-2">{{ $group->translate('name') }}</span>
```
Comment (glennjacobs) [inline packages/admin/resources/views/livewire/components/customers/show.blade.php:70]: Same here

File: packages/admin/src/Http/Livewire/Components/Settings/Taxes/AbstractTaxZone.php [hunk 2]
```diff
@@ -294,7 +294,7 @@ protected function syncCustomerGroups()
             });
 
             return [
-                'name' => $group->name,
+                'name' => $group->translate('name'),
```
Comment (glennjacobs) [inline packages/admin/src/Http/Livewire/Components/Settings/Taxes/AbstractTaxZone.php:297]: here

File: packages/admin/src/Http/Livewire/Dashboard.php [hunk 3]
```diff
@@ -315,7 +315,9 @@ public function getCustomerGroupOrdersProperty()
 
         $customerGroups = CustomerGroup::get();
 
-        $labels = $customerGroups->pluck('name')->toArray();
+        $labels = $customerGroups->map(
+            fn (CustomerGroup $group) => $group->translate('name')
```
Comment (glennjacobs) [inline packages/admin/src/Http/Livewire/Dashboard.php:319]: here

File: packages/core/src/Models/CustomerGroup.php [hunk 4]
```diff
@@ -6,15 +6,15 @@
 use Lunar\Base\BaseModel;
 use Lunar\Base\Traits\HasDefaultRecord;
 use Lunar\Base\Traits\HasMacros;
-use Lunar\Base\Traits\HasMedia;
+use Lunar\Base\Traits\HasTranslations;
```
Comment (glennjacobs) [inline packages/core/src/Models/CustomerGroup.php:9]: here

File: packages/core/src/Models/CustomerGroup.php [hunk 5]
```diff
@@ -6,15 +6,15 @@
...
 use Lunar\Database\Factories\CustomerGroupFactory;
 
 class CustomerGroup extends BaseModel
 {
     use HasFactory;
-    use HasMedia;
     use HasDefaultRecord;
     use HasMacros;
+    use HasTranslations;
```
Comment (glennjacobs) [inline packages/core/src/Models/CustomerGroup.php:17]: here

File: packages/admin/resources/lang/en/partials.php [hunk 6]
```diff
@@ -162,6 +162,10 @@
     'customers.purchase-history.revenue' => 'Revenue',
     'customers.purchase-history.order_count' => 'No. Orders',
     'customers.purchase-history.last_ordered' => 'Last Ordered',
+    /**
+     * Customer groups.
+     */
+    'customer-group.delete' => '',
```
Comment (glennjacobs) [inline packages/admin/resources/lang/en/partials.php:168]: Why would we want a blank lang file entry?

File: packages/admin/src/Http/Livewire/Components/Settings/Tables/CustomerGroupsTable.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    public bool $canSaveSearches = true;
+
+    /**
+     * {@inheritDoc}
+     */
+    protected $listeners = [
+        'saveSearch' => 'handleSaveSearch',
```
Comment (glennjacobs) [inline packages/admin/src/Http/Livewire/Components/Settings/Tables/CustomerGroupsTable.php:36]: Saved search code, but it's not being used? Does this matter?

**Review by glennjacobs — APPROVED** [review]

**Pull request merged**